multi/nrf91-multi/uart.c

```diff
@@ -0,0 +1,536 @@
+	uarte_stoptx,
+	uarte_flushrx = 11,
+	uarte_events_cts = 64,
+	uarte_events_ncts,
+	uarte_events_rxdrdy,
+	uarte_events_endrx = 68,
+	uarte_events_txdrdy = 71,
+	uarte_events_endtx,
+	uarte_events_error,
+	uarte_events_rxto = 81,
+	uarte_events_rxstarted = 83,
+	uarte_events_txstarted,
+	uarte_events_txstopped = 86,
+	uarte_inten = 192,
+	uarte_intenset,
+	uarte_intenclr,
+	uarte_errorsrc = 288,
+	uarte_enable = 320,
+	uarte_psel_rts = 322,
+	uarte_psel_txd,
+	uarte_psel_cts,
+	uarte_psel_rxd,
+	uarte_baudrate = 329,
+	uarte_rxd_ptr = 333,
+	uarte_rxd_maxcnt,
+	uarte_rxd_amount,
+	uarte_txd_ptr = 337,
+	uarte_txd_maxcnt,
+	uarte_txd_amount,
+	uarte_config = 347 };
+
+
+enum { no_parity,
+	parity_even
+};
+
+
+enum { baud_9600 = 0x00275000,
+	baud_115200 = 0x01d60000,
+	baud_230400 = 0x03b00000,
+	baud_250000 = 0x04000000,
+	baud_460800 = 0x07400000,
+	baud_921600 = 0x0f000000,
+	baud_1000000 = 0x10000000 };
+
+
+enum { uart0 = 0,
+	uart1,
+	uart2,
+	uart3 };
+
+
+/* indicates whether the tx line is ready to transmit new data */
+static bool uart_txready(uart_ctx_t *ctx)
+{
+	bool ret;
+	ret = ctx->init;
+
+	mutexLock(ctx->irqlock);
+	/* before the first transaction tx is also ready even if the event hasn't occurred */
+	if (ctx->init == false) {
+		/* true means that the data has been sent from txd */
+		ret = *(ctx->base + uarte_events_txdrdy);
+		*(ctx->base + uarte_events_txdrdy) = 0u;
+		ctx->init = false;
+	}
+	mutexUnlock(ctx->irqlock);
+
+	return ret;
+}
+
+
+static int uart_irqHandler(unsigned int n, void *arg)
+{
+	uart_ctx_t *ctx = (uart_ctx_t *)arg;
+
+	if (*(ctx->base + uarte_events_rxdrdy) != 0u) {
+		/* clear rxdrdy event flag */
+		*(ctx->base + uarte_events_rxdrdy) = 0u;
+		ctx->rxdCount++;
+		ctx->rxdrdy = true;
+	}
+
+	if (*(ctx->base + uarte_events_endtx) != 0u) {
+		/* disable endtx interrupt and clear flag */
+		*(ctx->base + uarte_events_endtx) = 0u;
+		*(ctx->base + uarte_intenclr) = 0x100u;
+	}
+
+	if (*(ctx->base + uarte_events_endrx) != 0u) {
+		/* clear endrx event flag */
+		*(ctx->base + uarte_events_endrx) = 0u;
+		ctx->resetCnt = true;
+		*(ctx->base + uarte_startrx) = 1u;
+	}
+
+	return 0;
+}
+
+
+static void uart_irqthread(void *arg)
+{
+	uart_ctx_t *ctx = (uart_ctx_t *)arg;
+	bool keptIdle = false;
+	int i = 0;
+
+	while (true) {
+		mutexLock(ctx->irqlock);
+		/* wait until new data is received and libtty fifo is not empty */
+		while (!ctx->rxdrdy && (libtty_txready(&ctx->ttyCommon) == 0)) {
+			condWait(ctx->cond, ctx->irqlock, 0);
+		}
+		mutexUnlock(ctx->irqlock);
+
+		ctx->rxdrdy = false;
+		dataBarier();
+
+		/* send the data directly from dma */
+		for (; (ctx->cnt < ctx->rxdCount) && (ctx->cnt < UART_RX_DMA_SIZE); ctx->cnt++) {
+			libtty_putchar(&ctx->ttyCommon, ctx->rxDma[ctx->cnt], NULL);
+		}
+
+		/* check whether all bytes from the previous transaction have been read */
+		if (ctx->resetCnt && (ctx->cnt >= UART_RX_DMA_SIZE)) {
+			ctx->cnt = 0;
+			ctx->rxdCount = ctx->rxdCount - UART_RX_DMA_SIZE;
```

Comment:
Please note that this operation is not atomic
- rxdCount value is taken here,
- interrupt is triggered, it increments and stores new rxdCount value,
- rxdCount is updated here, using old value.

I think it will cause one byte to be lost.

Or is it impossible to an intterupt be generated here?

Reply:
You are right, I've decided to use stdatomic